### ISSUE TITLE
Running windows game server

### ISSUE THREAD

**Issue by cyriltovena** [googleforgames/agones#54]
Game production usually works on windows first then port to linux eventually at the end of the development. Having windows support would help the adoption rate.

What does it takes to run a windows game server ?

- We should find a way to detect that the game server resource is windows ? May be a resource parameter ?
- Does windows support sidecar concept ? Apparently from [here at the bottom](http://blog.kubernetes.io/2017/09/windows-networking-at-parity-with-linux.html) it says windows support only one container per pod, but should work on windows server 1709. We should try it.
- If yes, create a sidecar using [windows nano](https://hub.docker.com/r/microsoft/nanoserver/).
- add a windows game server example.

Testing will be difficult as windows support is still in beta since k8s 1.5 but apparently greatly improve in 1.9.

Documentation :
https://kubernetes.io/docs/getting-started-guides/windows/
http://blog.kubernetes.io/2017/09/windows-networking-at-parity-with-linux.html
https://github.com/kubernetes/community/tree/master/sig-windows
https://docs.microsoft.com/en-us/windows-server/get-started/whats-new-in-windows-server-1709

**Comment by markmandel** [comment]
Today - there is an update!
https://kubernetes.io/blog/2019/03/25/kubernetes-1-14-release-announcement/

**Comment by Kuqd** [comment]
Kind issue to look at: https://github.com/kubernetes-sigs/kind/issues/410

**Comment by markmandel** [comment]
Big news!!!

https://cloud.google.com/kubernetes-engine/docs/release-notes?hl=en

> The ability to create clusters with node pools running Microsoft Windows Server is now in Beta. This feature is currently only available in the Rapid release channel.

**Comment by EricFortin** [comment]
I have taken a stab at this in the past few days. I somewhat succeeded and thought I would recap my work in this issue so that we may be able to make it happen.

**Disclaimer: I was mainly interested in seeing if this could even work. I hacked a few things and work around others.**

### Building sidecar and simple udp images for Windows

#### SDK

* I built the sdk binary from Agones build system for Windows with following command:  `make build-agones-sdk-binary-windows`. *I had to switch to Go to at least 1.13.6 because of a bug in Windows: https://github.com/golang/go/issues/35447*
* I then wrote a simple `Dockerfile` to build an image. I skipped some security(like creating a user) in order to go faster.

Here's an example of my final `Dockerfile`:

```
# Go runtime methods link to some Windows DLLs that are not in nanoserver. See https://github.com/StefanScherer/dockerfiles-windows/pull/197 for the workaround implemented here.
# Using multi stage build to copy it and still keep a relatively small image.
FROM mcr.microsoft.com/windows/servercore:1909 as core
WORKDIR /app

FROM mcr.microsoft.com/windows/nanoserver:1909

COPY --from=core /windows/system32/netapi32.dll /windows/system32/netapi32.dll

WORKDIR /app
COPY ./sdk-server.windows.amd64.exe /app/sdk-server.exe
ENTRYPOINT ["/app/sdk-server.exe"]
```
  
**Windows containers images need to be build on the same OS that they are targetting. When trying to build on my machine(1903) with a `FROM windows/nanoserver:1909` I had an error so I ultimately built my images on a VM.**

#### simple-udp

I used simple-udp to test as this is simple and can be cross-compiled. I used this command from Agones clone:

```
docker run --rm -e "GOOS=windows" -e "GOARCH=amd64" -e "GO111MODULE=on" -w /go/src/agones.dev/agones -v /f/go/src/agones.dev/agones/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.helm:/root/.helm -v /f/go/src/agones.dev/agones:/go/src/agones.dev/agones -v /f/go/src/agones.dev/agones/build//.gomod:/go/pkg/mod -v /f/go/src/agones.dev/agones/build//.gocache:/root/.cache/go-build agones-build:c16b1f68c7 go build -mod=vendor \
        -o /go/src/agones.dev/agones/examples/simple-udp/simple-udp.windows.amd64.exe  -ldflags "-X agones.dev/agones/pkg.Version=1.6.0-7988111" agones.dev/agones/examples/simple-udp
```

I then wrote a similar `Dockerfile` as the previous one to build the image.

### Cluster creation

I followed GKE's [documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/creating-a-cluster-windows#gcloud) to create a cluster to host Agones with Windows node. It seems we should be able to run in 1.15 but when I tried with this version, I couldn't add the Windows node pool so I reverted to use 1.16 which at the time of writing gave me a 1.16.8-gke15. I used those commands to create the cluster:

```
gcloud container clusters create special-cluster \
    --enable-ip-alias \
    --num-nodes=1 \
    --machine-type=n1-standard-4 \
    --cluster-version=1.16

gcloud container node-pools create win-node-pool \
    --cluster=special-cluster \
    --image-type=WINDOWS_SAC \
    --no-enable-autoupgrade \
    --metadata disable-legacy-endpoints=true \
    --machine-type=n1-standard-4 \
    --num-nodes=1

gcloud container node-pools create agones-system \
  --cluster=special-cluster \
  --no-enable-autoupgrade \
  --metadata disable-legacy-endpoints=true \
  --node-taints agones.dev/agones-system=true:NoExecute \
  --node-labels agones.dev/agones-system=true \
  --num-nodes=1 \
  --machine-type=n1-standard-4
```

### Agones installation

We need to replace the SDK image that default install use since it is the one build as a linux based container. Helm has the settings to overwrite. Unfortunately, Agones use a single repo url and only the name and tags can be replaced. We can then either host all images in an external repo and replace the image repo settings or we can cheat a little(which I did but it'd be better not to)

Here's how I installed Agones: `helm install --set agones.image.sdk.name=agones-sdk-win --set agones.image.sdk.tag=1909 --name my-release --namespace agones-system agones/agones`

This means the controller will insert sidecar with image at gcr.io/agones-images/agones-sdk-win:1909. On my side, since I built the image from the actual cluster's Windows node, I simply had to tag it like so. It worked in my favor since Kubernetes' default `PullPolicy` is `IfNotPresent`. This isn't great(it only works on single node node pool) and next time I will go with hosting them all in our repo.

### Deploying game servers

I used Agones' example and added a node selector to target Windows machine. Here's the content of my `fleet.yaml`:

```yaml
apiVersion: "agones.dev/v1"
kind: Fleet
metadata:
  name: simple-udp
spec:
  replicas: 2
  template:
    spec:
      ports:
      - name: default
        containerPort: 7654
      template:
        spec:
          nodeSelector:
            kubernetes.io/os: windows
          containers:
          - name: simple-udp
            image: <redacted>
            resources:
              requests:
                memory: "64Mi"
                cpu: "100m"
              limits:
                memory: "64Mi"
                cpu: "200m"
```

After that, I had 2 game servers running in my cluster. Allocation works too.

```
$k get gs
NAME                     STATE   ADDRESS         PORT   NODE                                              AGE
simple-udp-wk2rq-9r6b7   Ready   <redacted>      7081   gke-special-cluster-win-node-pool-568b60bb-dh14   13h
simple-udp-wk2rq-vqjvq   Ready   <redacted>      7145   gke-special-cluster-win-node-pool-568b60bb-dh14   21h
```

### End of good news

I was not able to contact my game server. So I did some investigation to diagnose: *Firewall rules is correctly setup as when running the sdk in local mode and simple-udp directly on Windows VM, I was able to contact the server.*

To run in GKE with Windows node, we have to enable ip alias which means each pod have their own IPs. Kubernetes then need to setup routing when using host ports. So I tested from within the cluster if I was able to reach game servers. Here are some results:

* From VM running pod, hitting externalIP:hostPort doesn't work (might be due to limitation of K8s Windows integration)
* From VM running pod, hitting localhost:hostPort doesn't work (might be due to limitation of K8s Windows integration)
* From VM running pod, hitting host internalIP:hostPort doesn't work
* From VM running pod, hitting podIP:containerPort works
* From pod in another VM(linux), hitting externalIP:hostPort doesn't work
* From pod in another VM(linux), hitting internalIP:hostPort doesn't work
* From pod in another VM(linux), hitting podIP:containerPort works

That means the game server hosting in Windows container is working but there are issues with the networking setup. After snooping around on the Windows VM, I found out that GKE uses `L2bridge` type network driver and specially [win-bridge] (https://github.com/containernetworking/plugins/tree/master/plugins/main/windows/win-bridge) CNI plugin. It looks like this plugin supports `hostPort` mapping where it would ask the Host Network Service(HNS) to configure some routing rules. At this point, my Windows Networking knowledge is coming a bit short and I can't really debug CNI integration in GKE.

I tested a standard cluster(only Container optimized OS VMs) and it worked so it point to a shortcoming in the Windows CNI implementation.

To be continued ...

**Comment by cyriltovena** [comment]
> Here's how I installed Agones: helm install --set agones.image.sdk.name=agones-sdk-win --set agones.image.sdk.tag=1909 --name my-release --namespace agones-system agones/agones

Would be great to be able to specify the SDK via an annotation, this way you don't need to reinstall agones and you can run different sdk version.

Nice experiment.

**Comment by roberthbailey** [comment]
> We need to replace the SDK image that default install use since it is the one build as a linux based container. Helm has the settings to overwrite. Unfortunately, Agones use a single repo url and only the name and tags can be replaced. We can then either host all images in an external repo and replace the image repo settings or we can cheat a little(which I did but it'd be better not to)

On GKE (I think this is also true for other providers) you have to have at least one node pool running linux VMs for system containers even if your are running your workloads on Windows. So you could just run the linux agones controllers next to the other linux Kubernetes controllers.

**Comment by roberthbailey** [comment]
I had been wondering about hostPort support. I read through the GKE and Kubernetes docs and didn't see a clear answer as to whether (or how well) they would work. Maybe someone on the GKE windows or GKE networking team can help. We will reach out internally.

**Comment by WeetA34** [comment]
Hello,
I've setup a local cluster on my windows workstation with 3 hyper-VM vm.
I deployed kub with rancher 2.4.4
1st try with flannel vxlan but i encounter a big which avoid windows container to reach kube api (10.43.0.1). It's a known issue which should have been fixed in kube 1.18.1.
So, i rebuilt the cluster with flannel l2bridge.
I'm stuck at the same point as Eric. The gameserver is running and is healthy but i'm not able to connect on the hostPort from another machine located on the same hyper-v network.
No issue with the simple-tcp linux.
I've checked firewall on the windows server core 1903 node. It seems ok.
I'm still investigating.

**Comment by WeetA34** [comment]
I ran almost all commands present in https://github.com/microsoft/SDN/tree/master/Kubernetes/windows/debug/ scripts
I didn't find the expected port

**Comment by WeetA34** [comment]
Same issue without agones, unable to connect to IIS container launched with the following command:
`kubectl run iis --namespace default --restart='Never' --port 80 --hostport 8080 --image=mcr.microsoft.com/windows/servercore/iis:windowsservercore-1903`

**Comment by markmandel** [comment]
Started a thread as well in #sig-windows on K8s slack:
https://kubernetes.slack.com/archives/C0SJ4AFB7/p1591381173181100

**Comment by markmandel** [comment]
Based on the conversation in the #sig-windows thread, I've filed/updated two tickets (see links above). and been told to check in in a week, to see if they can find some resources to work on hostPort support:

```
daschott  14 minutes ago
Yes, that would be a good issue to upvote, and include the information. I would considering raising a Windows issue on CNI plugins as well. Let me try to see internally if anyone on our team has cycles to pick this up. Please follow up again in a week if you haven't heard back by then.
```

**Comment by markmandel** [comment]
Possible good news!
https://github.com/containernetworking/plugins/releases/tag/v0.8.6 supports portMappings! So it may be a case of an updated CNI, and ensuring:

```
    "capabilities": {
        "dns": true,
        "portMappings":  true
    }
```

has been added to the CNI config.

**Comment by markmandel** [comment]
Looks like we do have it enabled on GCE:
https://github.com/kubernetes/kubernetes/blob/release-1.15/cluster/gce/windows/k8s-node-setup.psm1#L807

But need the CNI version is not up to 0.8.6 (which came out 23 days ago, so no surprise)

**Comment by WeetA34** [comment]
Hello Mark,
i already created an flannel image with 0.8.6 but i was not aware about portMappings.
i can now access iis when running iis windows container with the following command:
`kubectl run iis --namespace default --restart='Never' --port 80 --hostport 8000 --image=mcr.microsoft.com/windows/servercore/iis:windowsservercore-1903`
I have to try a game server now :)
Thank you Mark

**Comment by WeetA34** [comment]
Hello again
i just rebuilt the entire cluster with kubernetes 1.17.6
It works fine now :)
Thank you again Mark for pointing to the new CNi plugin portMappings attributes

```
$ kubectl get nodes -o wide
 NAME           STATUS   ROLES    AGE   VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                  KERNEL-VERSION       CONTAINER-RUNTIME 
kub2lin01      Ready    <none>   34m   v1.17.6   172.31.0.59     <none>        Ubuntu 18.04.4 LTS        4.15.0-101-generic   docker://19.3.11
 kub2master01   Ready    master   38m   v1.17.6   172.31.10.248   <none>        Ubuntu 18.04.4 LTS        4.15.0-101-generic   docker://19.3.11 
kub2win01      Ready    <none>   12m   v1.17.6   10.244.2.2      <none>        Windows Server Standard   10.0.18362.836       docker://19.3.5

$ kubectl get gs
 NAME                   STATE   ADDRESS      PORT   NODE        AGE 
simple-tcp-win-mpgsc   Ready   10.244.2.2   7612   kub2win01   3m56s

$ kubectl get pods -o wide 
NAME                   READY   STATUS    RESTARTS   AGE     IP           NODE        NOMINATED NODE   READINESS GATES 
simple-tcp-win-mpgsc   2/2     Running   5          4m24s   10.244.2.5   kub2win01   <none>           <none>

$  hostname
 alpine

$ ip addr show eth0 | grep 'inet '
     inet 172.31.8.45/20 scope global eth0

$ nc 172.31.10.83 7612 
helo toto
 ACK: helo toto
```

**Comment by WeetA34** [comment]
FYI, flannel windows daemonset pod doesn't restart after windows node reboot.
I created the following issue to kubernetes-sigs:
https://github.com/kubernetes-sigs/sig-windows-tools/issues/94

**Comment by WeetA34** [comment]
Successful test with a fleet:
```
kubemaster01$ kubectl create -f fleet.yaml
fleet.agones.dev/fleet-simple-tcp-win created

kubemaster01$ kubectl get gameservers
NAME                               STATE   ADDRESS      PORT   NODE        AGE
fleet-simple-tcp-win-c2bqh-wrt7c   Ready   10.244.2.2   7320   kub2win01   7s
fleet-simple-tcp-win-c2bqh-zz56w   Ready   10.244.2.2   7135   kub2win01   7s

kubemaster01$ kubectl scale fleet fleet-simple-tcp-win --replicas=10
fleet.agones.dev/fleet-simple-tcp-win scaled

kubemaster01$ kubectl get gameservers
NAME                               STATE   ADDRESS      PORT   NODE        AGE
fleet-simple-tcp-win-c2bqh-884h4   Ready   10.244.2.2   7303   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-8wcjj   Ready   10.244.2.2   7148   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-h4m4h   Ready   10.244.2.2   7123   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-knr9f   Ready   10.244.2.2   7011   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-sfzfk   Ready   10.244.2.2   7200   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-swss8   Ready   10.244.2.2   7105   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-wrt7c   Ready   10.244.2.2   7320   kub2win01   4m29s
fleet-simple-tcp-win-c2bqh-x67xv   Ready   10.244.2.2   7456   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-zbk7f   Ready   10.244.2.2   7855   kub2win01   3m29s
fleet-simple-tcp-win-c2bqh-zz56w   Ready   10.244.2.2   7135   kub2win01   4m29s

kubemaster01$ kubectl get gameservers -o jsonpath='{.items[*].spec.ports[0].hostPort}'
7303 7148 7123 7011 7200 7105 7320 7456 7855 7135

alpine$ for port in 7303 7148 7123 7011 7200 7105 7320 7456 7855 7135; do echo "hello gs listening on port ${port}" | nc -w 1 172.31.10.83 $port; done
ACK: hello gs listening on port 7303

ACK: hello gs listening on port 7148

ACK: hello gs listening on port 7123

ACK: hello gs listening on port 7011

ACK: hello gs listening on port 7200

ACK: hello gs listening on port 7105

ACK: hello gs listening on port 7320

ACK: hello gs listening on port 7456

ACK: hello gs listening on port 7855

ACK: hello gs listening on port 7135
```

**Comment by WeetA34** [comment]
Now i have to fix the problem of the wrong windows node ip.
I guess game clients will not be able to connect to game servers because the public IP of the windows node is not known on kubernetes.

**Comment by WeetA34** [comment]
Found it,
i added `--node-ip 172.31.10.83` in `C:\k\StartKubelet.ps1` and restarted kubelet service

**Comment by WeetA34** [comment]
Hello,

it works fine with UDP too
```
$ kubectl create -f fleet-udp.yaml
fleet.agones.dev/fleet-simple-udp-win created

$ kubectl get fleet
NAME                   SCHEDULING   DESIRED   CURRENT   ALLOCATED   READY   AGE
fleet-simple-udp-win   Packed       2         2         0           2       2m54s

$ kubectl get gs
NAME                               STATE   ADDRESS        PORT   NODE        AGE
fleet-simple-udp-win-rzpgd-jzfh6   Ready   172.31.10.83   7601   kub2win01   19s
fleet-simple-udp-win-rzpgd-xs68b   Ready   172.31.10.83   7445   kub2win01   19s

alpine:~$ for port in 7601 7445; do echo "hello gs listening on port ${port}" | nc -w 1 -u 172.31.10.83 $port; done
ACK: hello gs listening on port 7601
ACK: hello gs listening on port 7445
```

**Comment by WeetA34** [comment]
Hello of the day :)

I built a Windows Server core 1903 image with Xonotic Game server.
I successfully created a fleet with this image.
I was able to connect with the game to the Game server.
below some logs of the Game server
```
Server using port 26000
>>> Found 'listening' statement: 4
>>> Moving to READY: Server listening on address 0.0.0.0:26000
Server listening on address [0:0:0:0:0:0:0:0]:26000
Server listening on address 0.0.0.0:26000
Server listening on address [0:0:0:0:0:0:0:0]:26000
Error response from keygen server: Not allowed
Authenticated connection to 172.31.0.1:59722 has been established: client is glZQY................=@....., I am ow.....@.........
WeetA is connecting...
WeetA connected
WeetA connected
WeetA is now spectating
[BOT]Thunderstorm connected
[BOT]Discovery connected
[BOT]Mystery connected
[BOT]Delirium connected
unconnected changed name to [BOT]Thunderstorm
unconnected changed name to [BOT]Discovery
unconnected changed name to [BOT]Mystery
unconnected changed name to [BOT]Delirium
Receiving player stats failed: -404
[BOT]Discovery picked up Strength
[BOT]Discovery drew first blood!
[BOT]Delirium was gunned down by [BOT]Discovery's Shotgun
[BOT]Delirium got too close to [BOT]Discovery's Mortar grenade
[BOT]Thunderstorm slapped [BOT]Mystery around a bit with a large Shotgun
[BOT]Thunderstorm blew themself up with their own Mortar
[BOT]Discovery made a TRIPLE FRAG!
[BOT]Mystery was telefragged by [BOT]Discovery
[BOT]Discovery slapped [BOT]Delirium around a bit with a large Shotgun
[BOT]Discovery unlocked RAGE!
[BOT]Thunderstorm was gunned down by [BOT]Discovery's Shotgun
[BOT]Discovery slapped [BOT]Delirium around a bit with a large Shotgun
[BOT]Delirium ate [BOT]Mystery's rocket
[BOT]Thunderstorm was riddled full of holes by [BOT]Mystery's Machine Gun
[BOT]Mystery picked up Strength
[BOT]Mystery made a TRIPLE FRAG!
[BOT]Delirium was pummeled by a burst of [BOT]Mystery's Hagar rockets
[BOT]Delirium slapped [BOT]Thunderstorm around a bit with a large Shotgun
Client "WeetA" dropped
WeetA disconnected
Client "[BOT]Delirium" dropped
[BOT]Delirium disconnected
Client "[BOT]Thunderstorm" dropped
[BOT]Thunderstorm disconnected
Client "[BOT]Discovery" dropped
[BOT]Discovery disconnected
Client "[BOT]Mystery" dropped
[BOT]Mystery disconnected
```

**Comment by eddiebeazer** [comment]
> Here's how I installed Agones: helm install --set agones.image.sdk.name=agones-sdk-win --set agones.image.sdk.tag=1909 --name my-release --namespace agones-system agones/agones

I'm working on trying to replicate the same example since the server I'm using has to be ran on Windows.  I got everything working minus the sidecar.  How would I go about creating a 1909 version of the current 1.9 sdk?

**Comment by EricFortin** [comment]
@eddiebeazer Are you talking about the golang binary or the sdk image? What issues did you face with 1.9 particularly? I was able to just use the cross compiled sdk binary when I did my investigation.

**Comment by eddiebeazer** [comment]
> @eddiebeazer Are you talking about the golang binary or the sdk image? What issues did you face with 1.9 particularly? I was able to just use the cross compiled sdk binary when I did my investigation.

@EricFortin The sdk image.  I'm creating an unreal server with windows and I followed through the tutorials to adding the Agones BP to main map.  When I roll out my fleet config the server image is pulled fine but the Agones image doesn't pull at all when I try to use --set agones.image.sdk.name=agones-sdk-win --set agones.image.sdk.tag=1909. 

I'm completely new to Agones and haven't got the chance to dive deep yet in the sdk since I haven't been able to get a basic working example yet.  I don't know if I even need the sidecar if the sdk is implemented inside of my unreal server.

Overall what I'm trying to accomplish is to get Open Match, Agones and an Unreal Engine windows server working in GKE.  Currently got the server up fine and working on the Agones portion and got stuck with getting the sidecar to work correctly.

If it helps, I'm using the default terraform helm install from the agones repo.  I made a local copy of it so I could add these two statements to it

```terraform
set {
    name  = "agones.image.sdk.name"
    value = "agones-sdk-win"
  }

  set {
    name  = "agones.image.sdk.tag"
    value = "1909"
  }
```

Besides that, everything else is the same from the repo I copied it from.

**Comment by EricFortin** [comment]
Gotcha. Agones is not building Windows images at the moment and by default will pull from its own gcr repo. When I made my investigation, I had to build the image myself and I cheated as I built it directly on the node(and tagged it as if it was coming from the Agones repo) where the game server would be running because I didn't want to host it myself.

If you want to build your own image, I detailed the steps I took in my comments: https://github.com/googleforgames/agones/issues/54#issuecomment-637633950

**Comment by jeremyje** [comment]
To add on what https://github.com/googleforgames/agones/issues/54#issuecomment-637633950 says.

Given that Agones is a Linux centric project it's actually still possible to build Windows container images using [`DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build`](https://docs.docker.com/buildx/working-with-buildx/) as long as you're not using `RUN`. You need the latest version of docker for this since it requires buildkit. I can confirm this works for docker-in-docker scenarios since Agones does that from what I remember.

There's a Google Cloud Builder community project for building on Windows hosts https://github.com/GoogleCloudPlatform/cloud-builders-community/tree/master/windows-builder.

Lastly, for building Windows images on image hosts. It appears you can do this as long as the host is at or newer than the version under build though I wouldn't press my luck here since there are can be incompatibilities.

**Comment by WeetA34** [comment]
Hi,
i built a new test env on openstack with kubernetes 1.17.11, windows LTSC2019, agones 1.8.0 with flannel overlay (vxlan).
the sidecar container doesn't start workers
```
+ simple-tcp-win-l6qn6 › agones-gameserver-sidecar
+ simple-tcp-win-l6qn6 › simple-tcp-win
simple-tcp-win-l6qn6 simple-tcp-win 2020/09/04 12:47:11 Starting TCP server, listening on port 7654
simple-tcp-win-l6qn6 simple-tcp-win 2020/09/04 12:47:12 Creating SDK instance
simple-tcp-win-l6qn6 simple-tcp-win 2020/09/04 12:47:13 Starting Health Ping
simple-tcp-win-l6qn6 simple-tcp-win 2020/09/04 12:47:13 Marking this server as ready
simple-tcp-win-l6qn6 agones-gameserver-sidecar {"ctlConf":{"Address":"localhost","IsLocal":false,"LocalFile":"","Delay":0,"Timeout":0,"Test":"","TestSdkName":"","GRPCPort":9357,"HTTPPort":9358},"featureGates":"ContainerPortAllocation=true\u0026Example=true\u0026PlayerTracking=false\u0026SDKWatchSendOnExecute=false","message":"Starting sdk sidecar","severity":"info","source":"main","time":"2020-09-04T12:47:12.8533173Z","version":"1.8.0"}
simple-tcp-win-l6qn6 agones-gameserver-sidecar {"gsKey":"default/simple-tcp-win-l6qn6","message":"Created GameServer sidecar","severity":"info","source":"*sdkserver.SDKServer","time":"2020-09-04T12:47:13.1058722Z"}
simple-tcp-win-l6qn6 agones-gameserver-sidecar {"grpcEndpoint":"localhost:9357","message":"Starting SDKServer grpc service...","severity":"info","source":"main","time":"2020-09-04T12:47:13.1109151Z"}
simple-tcp-win-l6qn6 agones-gameserver-sidecar {"httpEndpoint":"localhost:9358","message":"Starting SDKServer grpc-gateway...","severity":"info","source":"main","time":"2020-09-04T12:47:13.1109151Z"}
simple-tcp-win-l6qn6 agones-gameserver-sidecar {"message":"shutting down sdk server","severity":"info","source":"main","time":"2020-09-04T12:47:22.8980415Z"}
simple-tcp-win-l6qn6 agones-gameserver-sidecar {"error":"failed to wait for caches to sync","message":"Could not run sidecar","severity":"fatal","source":"main","time":"2020-09-04T12:47:22.8980415Z"}
simple-tcp-win-l6qn6 simple-tcp-win 2020/09/04 12:47:23 Could not send health ping: could not send Health ping: EOF
```
any idea?
Thx

**Comment by markmandel** [comment]
`simple-tcp-win-l6qn6 agones-gameserver-sidecar {"error":"failed to wait for caches to sync","message":"Could not run sidecar","severity":"fatal","source":"main","time":"2020-09-04T12:47:22.8980415Z"}`

Sounds like there was an issue accessing the master? Or maaaaaybe an RBAC issue? i.e. it couldn't query the specific resource?

**Comment by jeremyje** [comment]
https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#network-modes
https://support.microsoft.com/en-us/help/4489899/windows-10-update-kb4489899 might not be installed.
Also some network modes are not available.

**Comment by WeetA34** [comment]
Yep,
it's a communication issue between windows pod and kubernetes Api.
i can contact kubernetes Api when i only have 1 master node and 1 windows worker node.
If i add 1 linux worker node, it doesn't work anymore

**Comment by WeetA34** [comment]
As communication with the kubernetes api server works only with master node public ip, i 'm using podpreset to overrides the two following environment variables:
- KUBERNETES_SERVICE_HOST
- KUBERNETES_SERVICE_PORT

```
apiVersion: settings.k8s.io/v1alpha1
kind: PodPreset
metadata:
  name: overrides-kubernetes-service-envs
spec:
  selector:
    matchLabels:
      agones.dev/role: gameserver
  env:
    - name: "KUBERNETES_SERVICE_HOST"
      value: "192.168.168.2"
    - name: "KUBERNETES_SERVICE_PORT"
      value: "6443"
```

**Comment by WeetA34** [comment]
I rebuilt everything with kubernetes 1.19.0
The kubernetes api connection issue from windows persists.
So i implemented the podpreset workaround i described in my previous message.
Now, the GS is ready but port Mapping doesn't work :(
I tested cniplugin 0.8.6 and 0.87 with portMappings capability set to true.

**Comment by WeetA34** [comment]
Hello @markmandel 
I'm testing Agones win in GKE.
I'm not able to connect to Win Game Servers (UDP or TCP)
I see win-bridge version is: CNI win-bridge plugin v0.8.2-gke.0
In cni config file, i see portMappings is set to True.
Firewall rules seem OK.
I'll check tomorrow if i forgot something.
Any idea?
Thank you

**Comment by markmandel** [comment]
How are you testing it?

**Comment by WeetA34** [comment]
With nc from remote.
I still have a firewall issue.
I enabled port 7000-8000 on the windows node itself but it doesn't
When i disable firewall profiles, i see a reset sent by the windows server

**Comment by markmandel** [comment]
> With nc from remote.

Sorry, I've no idea what this means

**Comment by WeetA34** [comment]
PortMappings should work on windows with your custom cni plugin, right?

**Comment by WeetA34** [comment]
netcat (nc)

**Comment by markmandel** [comment]
Ah gotcha - I mean, how are you testing the that windows binaries etc? You running a custom fork atm?

**Comment by markmandel** [comment]
Also, how are you configuring your firewall rules?

I assume it's the same as on linux nodes? set a network tag, then apply a VPC firewall rule?

**Comment by WeetA34** [comment]
Ha ok :)

I modified your terraform:
- enable vpc native 
- add another node pool for win nodes (image type WINDOWS_LTSC) with game-server-win network tag
- add game-server-win to target tags in the VPC firewall rule
- enable udp and tcp 7000-8000 on VPC firewall rule

then i manually built the following images directly on the windows node:
- agones-server-sdk:1.9.0
- simple-tcp-win
- simple-udp-win

**Comment by WeetA34** [comment]
FYI, i rebooted the windows node to see if something was not broken.
It's worse since :(
unable to get GS in ready state
```
{"message":"agones.dev/agones/pkg/client/informers/externalversions/factory.go:119: Failed to list *v1.GameServer: Get \"https://10.28.0.1:443/apis/agones.dev/v1/namespaces/default/gameservers?fieldSelector=metadata.name%3Dsimple-tcp-win-85qbd\u0026limit=500\u0026resourceVersion=0\": dial tcp 10.28.0.1:443: connectex: A socket operation was attempted to an unreachable network.","severity":"error","time":"2020-11-05T09:57:58.3491579Z"}
```

**Comment by WeetA34** [comment]
I rebuilt the windows node
i made a network trace.
I received syn packets from my workstation so firewall is OK.
But i don't see answer from pod through the hostport.
That's why i would know if cni plugin v0.8.2-gke.0 includes portMappings patch introduces in 0.86?
Thank you

**Comment by WeetA34** [comment]
I replaced v0.8.2-gke.0 versions of the following files by v0.8.7:
- C:\etc\kubernetes\cni\host-local.exe
- C:\etc\kubernetes\cni\win-bridge.exe
It works now.
```
$ kubectl get gs
NAME                   STATE   ADDRESS        PORT   NODE                                   AGE
simple-tcp-win-mb9nb   Ready   35.227.xx.yy   7210   gke-agones-win-win-bbdb8a25-bgc5       2m13s

$ echo "hello GS" | nc -w 2 35.227.xx.yy 7210
ACK: hello GS

```

**Comment by WeetA34** [comment]
So to be able to use Agones Windows on GKE:
- some few changes on the terraform (new node pool, enable ip alias/vpc native, adjust fw rules for tcp)
- add gcr.io/agones-images/agones-sdk:1.9.0 windows image
- replace host-local.exe and win-bridge.exe to v0.8.7

Dockerfile for gcr.io/agones-images/agones-sdk:1.9.0
```
FROM mcr.microsoft.com/windows/servercore:ltsc2019 as core
RUN powershell.exe -Command "curl.exe -k -LO https://github.com/googleforgames/agones/releases/download/v1.9.0/agonessdk-server-1.9.0.zip; Expand-Archive /app/agonessdk-server-1.9.0.zip; Move-Item agonessdk-server-1.9.0/sdk-server.windows.amd64.exe sdk-server.exe"

FROM mcr.microsoft.com/windows/nanoserver:1809
COPY --from=core /windows/system32/netapi32.dll /windows/system32/netapi32.dll
COPY --from=core /app/sdk-server.exe /app/sdk-server.exe
WORKDIR /app
ENTRYPOINT ["/app/sdk-server.exe"]
```

**Comment by markmandel** [comment]
@jeremyje - any thoughts? I've no idea about Windows and GKE.

**Comment by markmandel** [comment]
@WeetA34 glad you could get it working at least somehow! That's pretty awesome.

What version of GKE are you running against? Don't think I saw.

**Comment by jeremyje** [comment]
Also are you using WINDOWS_LTSC or WINDOWS_SAC for your nodepool?

**Comment by WeetA34** [comment]
@markmandel same version as your terraform module (1.16)
@jeremyje WINDOWS_LTSC
i opened a support case. They advised to create a custom image

**Comment by WeetA34** [comment]
not sure how to use it yet as i only see image_type in terraform resource
Will check later

**Comment by jeremyje** [comment]
[`gcloud container node-pools describe [NODE_POOL_NAME] --cluster [CLUSTER_NAME] --zone [ZONE_NAME]`](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/describe)

Or look up the node pool details in Cloud Console.

Terraform might tell you but that's going to be roundabout since it's commonly not explicit.

**Comment by daschott** [comment]
Just FYI
Re: overlay support for portMappings capabilities, I believe this should be enabled in `sdnoverlay` plugin here: https://github.com/microsoft/windows-container-networking

Keep in mind these plugins only work with containerD runtime:
https://github.com/kubernetes/enhancements/issues/1001

**Comment by WeetA34** [comment]
@jeremyje the node pool has an image_type attribute set to WINDOWS_LTSC.
The instance group and the instance template are automaticaly created so i don't see yet how to use my custom image.

**Comment by WeetA34** [comment]
@daschott
When i played with flannel vxlan on openstack VMS, portMappings didn't work.
GKE use win-bridge
I'll try later this sdnoverlay plugin on openstack again.
Thank you for the information

**Comment by jeremyje** [comment]
A few things here:

GKE Windows does not support containerd. The upstream Kubernetes support is in putting the finishing touches on that hopefully released in v1.20.

You cannot use custom images for GKE Windows. You can however use a startup script to change the configuration of the node. https://gist.github.com/jeremyje/523f57f844cfe8c604aa4dc2b394cf07 is an example of installing a Windows Update to a Windows SAC node and then manually configuring HNS a bit for Local IP preservation. Keep in mind the startup script needs to be idempotent between reboots. Windows Updates Service are disabled by default btw.

When it comes to GKE issues it's hard to give guidance on issues without knowing the version. Windows support is relatively new and features are being added quickly.

**Comment by WeetA34** [comment]
thank you @jeremyje 
Startup script is a good idea to perfom the following operations:
- update host-local.exe and win-bridge.exe to the proper version
- restart proper services
- build agones-sdk:1.9.0 image because it doesn't exist in gcr.io/agones-images/ for windows

**Comment by jeremyje** [comment]
You shouldn't need to build agones-sdk as part of a startup script flow. You'll want to do that out of node startup because that's an expensive operation. You can build that image and host in gcr.io or any docker image registry.

For the CNI try the following version or above:

- 1.16.8-gke.24 
- 1.16.9-gke.9
- 1.16.10-gke.2
- 1.17.5-gke.10
- 1.17.6-gke.8

According to the release channels (I'm going to assume your Terraform is using releaseChannel=STABLE) https://cloud.google.com/kubernetes-engine/docs/release-notes
The current version of a new cluster is 1.16.13-gke.401 which is sufficient to get the newer CNI v0.8.6-gke.0.

**Comment by WeetA34** [comment]
humm i built a cluster with 1.16.15-gke.3500. CNI is v0.8.2-gke.0
Regarding the image build, i have to check if it's possible to define the repository (or image url) of the agones-sdk sidecar container when we create a game server.

**Comment by jeremyje** [comment]
That's strange. I just created a cluster with a v1.16.15-gke.1600 (sorry slightly earlier version) and CNI 0.8.6 is installed.
I suspect your master is 1.16.15-gke.3500 but the Windows node pool is an earlier version please confirm.

2020/11/05 20:26:02 GCEMetadataScripts: windows-startup-script-ps1: Downloaded https://storage.googleapis.com/gke-release/kubernetes/release/v1.16.15-gke.1600/kubernetes-node-windows-amd64.tar.gz (SHA1 = b8bfdea71dadf199b4861cddf578171a51228fdd)
2020/11/05 20:25:57 GCEMetadataScripts: windows-startup-script-ps1: Setting environment variable: WINDOWS_CNI_VERSION = v0.8.6-gke.0

If I read your concern about the image correctly I think your concern is that that the agones sdk side cars are hard coded so you need to build the image on the host with the image tag it expects as a workaround for this limitation. If that's the case may I suggest this then:
1. Build the image once out of band (CI or on your desktop for testing).
1. Push that image to your registry of choice.
1. Add a startup script that does the following: (in powershell)

```powershell
& docker pull $YOUR_IMAGE
& docker tag -t $YOUR_IMAGE gcr.io/agones-images/agones-sdk:1.9.0
```

I'm not sure if it's possible but you may need to change the sidecar's image pull policy to IfNotPresent because since the tag is different it may complain that the upstream image manifest is different.

Lastly, I looked at gcr.io/agones-images/agones-sdk:1.9.0. It's a straight linux amd64 image. Agones should look building a multi-arch image to support Windows. If you're on Docker 19.03 (I think?) then you can use `DOCKER_CLI_EXPERIMENTAL=enabled docker build buildx` tag each OS target flavor with a `-linux_amd64` `-windows_amd64_ltsc2019` etc and then you can use `DOCKER_CLI_EXPERIMENTAL=enabled docker manifest create gcr.io/agones-images/agones-sdk:$NEXT_VERSION [each os specific tag...]` to have a unified reference. This workflow does not require Windows to do. For a makefile example see csi-release-tools, https://github.com/kubernetes-csi/csi-release-tools/blob/master/build.make#L144. A word of caution for buildx is that the os.version is not embedded. Example workaround: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/docker/build.sh#L103. I believe the next version of docker fixes this, https://github.com/docker/cli/pull/2578.

**Comment by WeetA34** [comment]
I've destroyed the cluster so i can't check right now. i'll rebuild it tomorrow to verify.

You're right, docker pull & tag is definitely a better option.
It's already IfNotPresent because i had not issue during my test with local docker build. When i deployed gameservers, the sidecar container didn't try to pull the image.

The pull & tag is a temporary workaround. I guess Agones will build multi arch in the future if they find an interest of having windows game servers :)

Thank you @jeremyje for your advices

**Comment by WeetA34** [comment]
Hello @jeremyje 

There is a mismatch between kube-env custom metadata WINDOWS_CNI_VERSION environment variable and installed binaries:
- 1.16.13-gke.403: v0.8.6-gke.0
```
PS C:\> C:\etc\kubernetes\node\bin\kubelet.exe --version
Kubernetes v1.16.13-gke.403

PS C:\> $env:WINDOWS_CNI_VERSION
v0.8.6-gke.0

PS C:\> C:\etc\kubernetes\cni\win-bridge.exe
CNI win-bridge plugin v0.8.2-gke.0
```
- v1.16.15-gke.1600: v0.8.6-gke.0
```
PS C:\> C:\etc\kubernetes\node\bin\kubelet.exe --version
Kubernetes v1.16.15-gke.3500

PS C:\> $env:WINDOWS_CNI_VERSION
v0.8.6-gke.0

PS C:\> C:\etc\kubernetes\cni\win-bridge.exe
CNI win-bridge plugin v0.8.2-gke.0
```

**Comment by WeetA34** [comment]
Hi @jeremyje
i tried to use windows-startup-script-ps1 metadata key with a copy of the original script i modified for my usecase.
it doesn't work. i get the following error
```
Error: error creating NodePool: googleapi: Error 400: Cluster.node_config.metadata.key cannot be the reserved key "windows-startup-script-ps1"., badRequest
```
So i understand i can only use windows-startup-script-url metadata key like in your script, right?
I guess it's executed after the windows-startup-script-ps1, correct?

Thank you

**Comment by markmandel** [comment]
@WeetA34 would be awesome if you could - would love to see a doc that outlines how you made Agones work on Windows, and it might be something we can replicate, or other people can try as well.

Get some more eyes on the details!

**Comment by jeremyje** [comment]
@WeetA34 
Yes you should use the `windows-startup-script-url` field. The metadata fields that are reserved are listed here, https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/NodeConfig (see the metadata field).
The detailed docs for startup scripts are here: https://cloud.google.com/compute/docs/startupscript#providing_a_startup_script_for_windows_instances.

The easiest way is to store the ps1 file in GCS, the file can be private if the service account of the machine is authorized to have read access on the file.

The other startup script fields are reserved by Kubernetes because they run all the node setup logic on the guest to ultimately join the cluster.

The startup script specified by URL will run after the node setup logic on every boot. You should make the script idempotent for this reason.

**Comment by WeetA34** [comment]
Hi @jeremyje 
Thank you for the confirmation.
For now, i created a gist like you did which:
- update cni plugin if it’s not v0.8.6-gke.0
- build the agones-sdk:1.9.0 local image if it doesn’t exist (build is fast)

**Comment by WeetA34** [comment]
Hi @markmandel,
I have to make a clean pull request of the terraform module. 
Is-it possible to build agones-sdk images for windows and add them in the registry?
So, i’ll not need to build (or pull/tag) local image.
Thx

**Comment by markmandel** [comment]
@WeetA34 to be honest, I've no idea where to start with building windows images -- so having a step by step guide to follow would be very useful.

**Comment by jeremyje** [comment]
Windows Node Pool Terraform is documented: https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/using_gke_with_terraform#windows-node-pools
And there's more information here: https://cloud.google.com/kubernetes-engine/docs/how-to/creating-a-cluster-windows.
If there's any gaps or things that are unclear please reach out.
@markmandel See https://github.com/googleforgames/agones/issues/54#issuecomment-722641544 for a bit of guidance on building Windows images for Agones given the current workflow. I'd start with `:ltsc2019` and build up from there.

**Comment by WeetA34** [comment]
@jeremyje 
I mean propose a pull request for agones cluster module

**Comment by WeetA34** [comment]
@markmandel 
@jeremyje sent some infos about multi-arch build in a previous comment

**Comment by jeremyje** [comment]
Confirmed the bug on CNI version (sorry for the confusion) add this to your startup script, https://gist.github.com/jeremyje/c97d43506a47181028ebeacc41dcf256.
GKE 1.17 the new version is used.


> Hello Jeremy Edwards
> 
> There is a mismatch between kube-env custom metadata WINDOWS_CNI_VERSION environment variable and installed binaries:
> 
> * 1.16.13-gke.403: v0.8.6-gke.0
> 
> ```
> PS C:\> C:\etc\kubernetes\node\bin\kubelet.exe --version
> Kubernetes v1.16.13-gke.403
> 
> PS C:\> $env:WINDOWS_CNI_VERSION
> v0.8.6-gke.0
> 
> PS C:\> C:\etc\kubernetes\cni\win-bridge.exe
> CNI win-bridge plugin v0.8.2-gke.0
> ```
> 
> * v1.16.15-gke.3500: v0.8.6-gke.0
> 
> ```
> PS C:\> C:\etc\kubernetes\node\bin\kubelet.exe --version
> Kubernetes v1.16.15-gke.3500
> 
> PS C:\> $env:WINDOWS_CNI_VERSION
> v0.8.6-gke.0
> 
> PS C:\> C:\etc\kubernetes\cni\win-bridge.exe
> CNI win-bridge plugin v0.8.2-gke.0
> ```
> 
> edit 1:
> I guess v0.8.2-gke.0 is already present in the image. That's why it's not replaced by the expected version
> 
> ```
> Info
> 2020-11-06 10:04:36.860 CET
> 2020/11/06 09:04:36 GCEMetadataScripts: windows-startup-script-ps1: Directory: C:\\etc\\kubernetes\\cni\r\n
> Info
> 2020-11-06 10:05:39.508 CET
> "windows-startup-script-ps1: Skip: C:\etc\kubernetes\cni\win-bridge.exe already exists, not overwriting it"
> Info
> 2020-11-06 10:05:39.524 CET
> "windows-startup-script-ps1: Skip: C:\etc\kubernetes\cni\host-local.exe already exists, not overwriting it"
> ```
> 
> Function Configure-CniNetworking from k8s-node-setup-psm1
> 
> ```
> function Configure-CniNetworking {
>   $CNI_RELEASE_VERSION = 'v0.8.2-gke.0'
>   if ((ShouldWrite-File ${env:CNI_DIR}\win-bridge.exe) -or
>       (ShouldWrite-File ${env:CNI_DIR}\host-local.exe)) {
>     $tmp_dir = 'C:\cni_tmp'
>     New-Item $tmp_dir -ItemType 'directory' -Force | Out-Null
> 
>     $release_url = (${env:WINDOWS_CNI_STORAGE_PATH} + '/' + ${env:WINDOWS_CNI_VERSION} + '/')
>     $tgz_url = ($release_url +
>         "cni-plugins-windows-amd64-${env:WINDOWS_CNI_VERSION}.tgz")
>     $sha_url = ($tgz_url + ".sha1")
>     MustDownload-File -URLs $sha_url -OutFile $tmp_dir\cni-plugins.sha1
>     $sha1_val = ($(Get-Content $tmp_dir\cni-plugins.sha1) -split ' ',2)[0]
>     MustDownload-File `
>         -URLs $tgz_url `
>         -OutFile $tmp_dir\cni-plugins.tgz `
>         -Hash $sha1_val
> 
>     Push-Location $tmp_dir
>     # tar can only extract in the current directory.
>     tar -xvf $tmp_dir\cni-plugins.tgz
>     Move-Item -Force host-local.exe ${env:CNI_DIR}\
>     Move-Item -Force win-bridge.exe ${env:CNI_DIR}\
>     Pop-Location
>     Remove-Item -Force -Recurse $tmp_dir
>   }
>   ...
> ```

**Comment by WeetA34** [comment]
Hello @jeremyje 
Yep. Google support has confirmed the issue in the case i opened. 
You should not use restart-computer for now because there is another issue for which i opened a case too: pods are not able to communicate with service network after reboot (dns resolution failure, kubernetes api unreachable, ...)
So for now, my startup script updates the cni plugin binaries and restarts kubelet and kube-proxy services. It works fine since i implemented this on my lab.
Thank you for the confirmation and for the gist.

**Comment by WeetA34** [comment]
Hi @jeremyje 
I moved to 1.17 as you said cni plugin version is the good one (0.8.6-gke.0). 
The network issue between pod and service subnet after worker node reboot is also resolved with 1.17.
Thank you

**Comment by markmandel** [comment]
Currently working on officially supporting 1.17 right now - so this sounds like good timing!

**Comment by WeetA34** [comment]
Yep because GKE 1.17 seems to be windows friendly :)
For now, it works fine for me with 1.17 with my own registry.
No need to use custom windows startup script.
I still have a "blocker" in my case for which i just opened the following feature request: https://github.com/googleforgames/agones/issues/1911

**Comment by jeremyje** [comment]
Sounds good. Sorry for all the issues on the GKE Windows side, it's pretty new and there are gaps but if there are any significant ones it'd be good to hear. Lastly, can you review the https://github.com/googleforgames/agones/pull/1894 to see if it fits your workflow. There's 1 more pending change I need to push which is to change the build to `make WITH_WINDOWS=1 [TARGET]`. Any general feedback about Windows on GKE is appreciated as well.

**Comment by markmandel** [comment]
#1894 is now merged! (See PR for details)

TLDR: there is a new optional `WITH_WINDOWS` param for building and pushing the appropriate Agones images to registries.

Apparently everything works best on Kubernetes 1.17 (or at least GKE 1.17) - which is where the next release will be supporting.

**Comment by josephbmanley** [comment]
Hey there! 👋 

So, I've gone through and checked on the current state of Windows and from my tests so far it has been working great! I think to have windows support more accessible I think the following needs to be done:
- [ ] Add Windows containers to automated build system (Currently Windows SDK image is not in the registry)
- [ ] Either add windows option to helm or add to docs to set agones.image.sdk.tag to the windows tag
- [ ] Setup default node selectors in helm charts to deploy on Linux

**Comment by TBBle** [comment]
If the Agones SDK server container image is pushed to different tags for different platforms/architectures, then a Manifest List can be pushed that references them, and the container runtime will pull the right image for its platform.

Then `agones.image.sdk.tag` can reference that manifest list by default, and is platform-agnostic.

**Comment by jeremyje** [comment]
https://github.com/googleforgames/agones/pull/1894 does the manifest magic already it just needs to be turned on in CI. At the time I cautioned against enabling it by default because it significantly increases the build time and I didn't want to complicate an Agones release. Since there's a release cut after the PR went in it's appropriate to turn on `WITH_WINDOWS` by default.

Beyond that like @josephbmanley said, all you pretty much need is some changes to Terraform to provision a [Windows Node Pool](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/using_gke_with_terraform#windows-node-pools) via conditional and the Helm templates need to be updated to have a parameter that basically inserts the [nodeSelector for Windows](https://cloud.google.com/kubernetes-engine/docs/how-to/deploying-windows-app#create_a_deployment_manifest_file).

/cc @markmandel 

There's 1 big caveat in that only Windows LTSC 2019 is supported (default for Kubernetes). Using it on 2004 and 20H2 will break in weird ways (this is likely going to be the version your dev box is on unless you run with Windows Server). Agones will need to update the CI image to install Docker Client 20.10 which comes with the os.version fix to `docker manifest amend`. This can be done today since the features accessed are client side and I've confirmed in a separate project it works. It's possible to get this behavior by sed grepping the manifest files locally but it's awful and to me more risky than just upgrading the client at this point.

**Comment by markmandel** [comment]
@josephbmanley 

Thanks for digging into this work!

> Setup default node selectors in helm charts to deploy on Linux 

I *think* we have this already? See [this documentation](https://agones.dev/site/docs/installation/creating-cluster/gke/#creating-the-cluster). Will that suffice?

> Either add windows option to helm or add to docs to set agones.image.sdk.tag to the windows tag

I don't think this is necessary (I'm fairly sure this is covered above), as with the way the manifest/registry operates, it can select the appropriate OS image as needed without having to be specific about it.

One thing to also add to the list - make sure the windows images are built and pushed as part of the release:
https://github.com/googleforgames/agones/blob/master/build/includes/release.mk#L74-L80

We should probably also add a section for documentation as well - likely a "Windows containers" page would be useful - maybe under `Installation > Create Cluster` Or maybe within each provider? Maybe also a "Guide" of some kind?

@jeremyje 

>  At the time I cautioned against enabling it by default because it significantly increases the build time

How significantly are we talking here? minutes, hours, days? :smile: 

> Agones will need to update the CI image to install Docker Client 20.10

We use whatever version comes with Cloud Build, also whatever we have on our work machines - sounds like we'll need to wait on supporting 2004 and 20H2 until the newer Docker version propagates out to those systems (if they haven't already).

Does that all make sense?

**Comment by TBBle** [comment]
> > Setup default node selectors in helm charts to deploy on Linux
> 
> I _think_ we have this already? See [this documentation](https://agones.dev/site/docs/installation/creating-cluster/gke/#creating-the-cluster). Will that suffice?

I'm not sure what your link is referring to (the `agones.dev/agones-system` taints, I guess?), but I believe @josephbmanley meant hardcoding something like
```
kubernetes.io/os: linux
kubernetes.io/arch: amd64
```
in the [`nodeSelector` for the controller](https://github.com/googleforgames/agones/blob/24902bf0abe4085beac1041f1dc0c7be25c955a8/install/helm/agones/templates/controller.yaml#L61-L64) to match the generated container images platform(s),, so they are not scheduled on Windows nodes (or ARM64 Linux nodes)

Users can also do this themselves using Helm values, but if the container os/arch is known statically, it seems nicer to make it part of the chart, than make it part of the config where it might be forgotten or overwritten.

Another option, the "legacy" option from the [Kubernetes docs](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) is to taint all your Windows nodes, and include tolerations on the Fleet's PodTemplates.

**Comment by markmandel** [comment]
Yeah exactly, the section that reads:

```
gcloud container node-pools create agones-system \
  --cluster=[CLUSTER_NAME] \
  --no-enable-autoupgrade \
  --node-taints agones.dev/agones-system=true:NoExecute \ # < this bit
  --node-labels agones.dev/agones-system=true \ # < and this bit
  --num-nodes=1
```

But I didn't realise you could do:

```
kubernetes.io/os: linux
kubernetes.io/arch: amd64
```

Which makes total sense - and should be backward compatible :+1: 

Do we need to something similar for the GameServer windows images? I assume we do?

**Comment by TBBle** [comment]
Ideally, `nodeSelectors` should be used in all Deployments to match the os and architectures of the manifest (or manifest list, where used). Since Windows containers are tied to the specific version of the kernel needed, that's also something that should be in the manifest lists based on the targets being built.

Sadly, k8s has been practically mono-os and mono-arch long enough most people and Helm charts aren't including appropriate `nodeSelectors` in their rollouts. Both Windows and ARM64 are starting to be serious k8s things, and there's [other surprise platforms around](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#multi-platform) that _someone_ must care about. I doubt I'll ever see a `GameServer` on a s390x Kubernetes node.

For illustration, once `WITH_WINDOWS` is turned on, the [current gameserver sidecar manifest list](https://github.com/googleforgames/agones/blob/24902bf0abe4085beac1041f1dc0c7be25c955a8/build/Makefile#L461-L462) implies that the `GameServer`-owned `Pod` would need something like this to prevent being scheduled on a node that couldn't run sdk-server container.

```yaml
spec:
  affinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      # Option 1: Linux && AMD64
      - matchExpressions:
        - key: kubernetes.io/os:
          operator: In
          values: [ linux ]
        - key: kubernetes.io/arch:
          operator: In
          values: [ amd64 ]
      # Option 2: Windows && AMD64 && [ ltsc2019 ]
      - matchExpressions:
        - key: kubernetes.io/os:
          operator: In
          values: [ windows ]
        - key: kubernetes.io/arch:
          operator: In
          values: [ amd64 ]
        - key: node.kubernetes.io/windows-build:
          operator: In
          values: [ "10.0.17763" ]
```

I expect that the above would never actually be used as-is. The main container in the `GameServer`'s `Pod` is likely to have more-specific architecture support (I'd be surprised if any Agones user is deploying manifest lists for their own game server container image, as differences in performance profile seem very likely), and also more-specific user-custom `nodeSelector` or `affinity` terms, to select the node-group with, e.g., appropriately-sized nodes and public-routable IP addresses so `hostPort` works.

I think in the end it'd be easier on everyone to simply build a wider variety of sdk-server image arch/os/version targets, so that GameServers are always constrained by the user, not the sdk-server container availability. ^_^

**Comment by josephbmanley** [comment]
@markmandel 

> > Either add windows option to helm or add to docs to set agones.image.sdk.tag to the windows tag
> 
> I don't think this is necessary (I'm fairly sure this is covered above), as with the way the manifest/registry operates, it can select the appropriate OS image as needed without having to be specific about it.
> 
> One thing to also add to the list - make sure the windows images are built and pushed as part of the release:
> https://github.com/googleforgames/agones/blob/master/build/includes/release.mk#L74-L80
> 
> We should probably also add a section for documentation as well - likely a "Windows containers" page would be useful - maybe under `Installation > Create Cluster` Or maybe within each provider? Maybe also a "Guide" of some kind?
> 
> @jeremyje

If it's true that it already selects the proper image based on the manifest, the default tags should be documented, as I had to specify the tag manually when using my own registry.

> > Agones will need to update the CI image to install Docker Client 20.10
> 
> We use whatever version comes with Cloud Build, also whatever we have on our work machines - sounds like we'll need to wait on supporting 2004 and 20H2 until the newer Docker version propagates out to those systems (if they haven't already).

Azure also only supports the current version as well. It seems all cloud providers are a bit behind on this.

**Comment by markmandel** [comment]
> I think in the end it'd be easier on everyone to simply build a wider variety of sdk-server image arch/os/version targets, so that GameServers are always constrained by the user, not the sdk-server container availability. ^_^

Sounds like what i think we're saying here is - put os specific constraints on the linux bits, leave GameServers as a exercise for the user on how to restrict game servers to their appropriate OS (in this case Windows) basically so they can just do what they want (although this sounds like something we should definitely document), and we don't get in their way.

Makes sense to me if that's the general consensus.

This multi-arch stuff make my head hurt :face_with_head_bandage:

**Comment by jeremyje** [comment]
The way the `make WITH_WINDOWS=1 build-images` is setup is that it'll [create multi-arch images between Linux and Windows](https://github.com/googleforgames/agones/blob/master/build/Makefile#L461). Meaning at the end of the day none of the references change. There's definitely complexity in managing the different manifests but it's 80% implemented already. The remaining (hard) 20% is managing different versions of Windows which is much lower priority. Generally speaking ltsc2019 is good enough and when ltsc2021 comes out will be the time to work the multi-arch.

Docker Client 20.10 is compatible with Docker Server 19.03. I've verified I can build like this in a different project.

> How significantly are we talking here? minutes, hours, days? 😄
+10-30 minutes based on machine type, hdd vs ssd, of the machine. If you parallelize you'll need to create a separate buildx context for each thread.

If there's only a sidecar I don't think there's any specific changes necessary. It'll run on Windows and Linux because the manifest makes that work. The pod container spec simply needs the `nodeSelector` to tell it to run on Windows and it should mostly just work. If there's containers that need to run on the Windows host outside of the sidecar then you'd need to basically add tolerations for Windows but I strongly suggest using the manifest based tagging for cross-platform workloads instead of using YAML to manage this.

**Comment by TBBle** [comment]
The issues I'm referring to are (pods being assigned to Windows nodes that don't have Windows containers images available) are prevented by [the example taint in the docs](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host), but I consider tainting Windows nodes to be a short-term workaround until all the things running on the cluster have appropriate nodeSelectors to ensure their container image manifests match the node arch/os/buildver.

Otherwise we'll go through this all again as ARM CPUs become more popular in k8s deployments.

**Comment by markmandel** [comment]
Quick thought I had on e2e testing:

Figured we could add some windows nodes to our e2e cluster, and then adjust some of our e2e tests to run both on Linux and Windows, by running the same test just with the windows node selector in the Pod template for the windows test.

**Comment by dzmitry-lahoda** [comment]
one last request - publish `simple-game-server-windows` to gcr:)

**Comment by roberthbailey** [comment]
If you pull `gcr.io/agones-images/simple-game-server:0.3` it will pull the windows version if you are on a windows node. That image is a manifest that contains references to the linux and windows images to pull. The windows image (if you wanted to pull it directly) is `gcr.io/agones-images/simple-game-server:0.3-windows_amd64-ltsc2019`.

**Comment by dzmitry-lahoda** [comment]
@roberthbailey i see that stuff works for simple-gs, but not for xonotic. i see xonotic has windows, but pulling it directly or with windows suffix fails on windows pod.

**Comment by TBBle** [comment]
Looking at https://console.cloud.google.com/gcr/images/agones-images/GLOBAL/xonotic-example?gcrImageListsize=30, it doesn't look like a new Xonitic manifest has been pushed since early 2020, and the Windows support was added in December 2020 in #1894.

So I don't think a Windows build of Xonotic has been pushed to the Agones repo, you probably have to build it yourself.

The [xonotic Makefile](https://github.com/googleforgames/agones/blob/release-1.15.0/examples/xonotic/Makefile) used by the cloudbuild scripts doesn't build a Windows image, so even if it had been built since Windows support was added, it wouldn't have been pushed. You can contrast with with [simepl-game-server's Makefile](https://github.com/googleforgames/agones/blob/release-1.15.0/examples/simple-game-server/Makefile) to see how much is missing, both building Windows images, and pushing the manifest list that lets you use one repository name from multiple platforms.

https://github.com/googleforgames/agones/pull/1894#issuecomment-737337635 suggests that the xonotic Windows image isn't cross-buildable from Linux, which might have been (or still be) a blocker for Agones's build pipeline. I'm not sure why else it didn't get the Windows support added to its Makefile when it was added to the Dockerfile.

So either an oversight, or a difficult problem to solve, I guess.

**Comment by TBBle** [comment]
Since I was looking at the simple-game-server's Makefile, [this TODO](https://github.com/googleforgames/agones/blob/release-1.15.0/examples/simple-game-server/Makefile#L64-L67) (and a couple of others in thsi file) are now completable, as Docker 20.10 has been [available on Cloud Build since May 2021](https://cloud.google.com/build/release-notes). I guess those TODOs might have been replicated in other Makefiles, but haven't looked.

**Comment by roberthbailey** [comment]
@jeremyje - do you have any cycles to look into making our xonotic example run on windows servers?

**Comment by jeremyje** [comment]
@TBBle Yes GCB has supported docker 20.10 for a while I'd strongly recommend upgrading to it because 19.03 is going end of life later this month, https://endoflife.software/applications/virtualization/docker-daemon.

@roberthbailey I unfortunately do not have cycles to work on this. I looked at it briefly here's the state. The `Dockerfile.windows` will only run on a Windows host because it's executing `RUN` commands. It can be adjusted to run on Linux but the `Makefile` will need to do the coordination of downloading the Xonotic zip file and extracting the files locally before copying them into the image via `COPY`. After that you can execute it as a `docker buildx build`. There's another issue where the `main.go` stdout scrubber cannot execute Xonotic while in the container. I'm not sure why but the first step would be to execute it again locally to see why it's breaking. If it isn't the next step I'd do is attempt to use the full windows docker image, `mcr.microsoft.com/windows:1809` to see if the repros. Xonotic is very sensitive to file structure and paths so it makes it difficult to make it run.

**Comment by markmandel** [comment]
@zmerlynn , @gongmax 🤔 I think the only thing left here is to setup some automated testing of a (?) windows cluster, probably on at least a single version of supported GKE?

Looks like there is some cleanup to do here as well with the build system, and hopefully there isn't much else? But I also say that as someone who isn't familiar with the windows container ecosystem either.

Marking as `awaiting-maintainer`, since this work is ongoing.
